Zend/zend_opcode.c

```diff
@@ -840,6 +839,198 @@ int pass_two_wrapper(zval *el)
+		 && cur_op->opcode != ZEND_JMPZ_EX
+		 && cur_op->opcode != ZEND_JMPNZ_EX
+		 && (cur_op->opcode != ZEND_QM_ASSIGN || (cur_op + 1)->opcode != ZEND_JMP) /* these two consecutive ops appear on ternary; the result of true branch is undefined for false branch */
+		 && cur_op->opcode != ZEND_CASE           /* exception for opcache, it might nowhere use the temporary (anyway bool, so no need to free) */
+		 && cur_op->opcode != ZEND_ROPE_ADD       /* the following opcodes reuse TMP created before */
+		 && cur_op->opcode != ZEND_ADD_ARRAY_ELEMENT
+		 && cur_op->opcode != ZEND_FAST_CALL      /* passes fast_call */
+		 && cur_op->opcode != ZEND_FETCH_CLASS    /* the following opcodes pass class_entry */
+		 && cur_op->opcode != ZEND_DECLARE_CLASS
+		 && cur_op->opcode != ZEND_DECLARE_INHERITED_CLASS
+		 && cur_op->opcode != ZEND_DECLARE_INHERITED_CLASS_DELAYED
+		 && cur_op->opcode != ZEND_DECLARE_ANON_CLASS
+		 && cur_op->opcode != ZEND_DECLARE_ANON_INHERITED_CLASS) {
+			if (done_pass_two) {
+				var = EX_VAR_TO_NUM(cur_op->result.var) - op_array->last_var;
+			} else {
+				var = cur_op->result.var;
+			}
+			/* Objects created via ZEND_NEW are only fully initialized after the DO_FCALL (constructor call) */
+			if (cur_op->opcode == ZEND_NEW) {
+				Tstart[var] = cur_op->op2.opline_num - 1;
+			} else {
+				Tstart[var] = cur_op - op_array->opcodes;
+			}
+		}
+		if (cur_op->op1_type & (IS_VAR | IS_TMP_VAR)) {
+			if (done_pass_two) {
+				var = EX_VAR_TO_NUM(cur_op->op1.var) - op_array->last_var;
+			} else {
+				var = cur_op->op1.var;
+			}
+			if (Tstart[var] != (uint32_t)-1
+			 && cur_op->opcode != ZEND_ROPE_ADD /* the following opcodes don't free TMP */
+			 && cur_op->opcode != ZEND_FETCH_LIST
+			 && cur_op->opcode != ZEND_CASE
+			 && cur_op->opcode != ZEND_FE_FETCH_R
+			 && cur_op->opcode != ZEND_FE_FETCH_RW) {
+				op_live_total += liveliness_kill_var(op_array, cur_op, var, Tstart, opTs);
+			}
+		}
+		if (cur_op->op2_type & (IS_VAR | IS_TMP_VAR)) {
+			if (done_pass_two) {
+				var = EX_VAR_TO_NUM(cur_op->op2.var) - op_array->last_var;
+			} else {
+				var = cur_op->op2.var;
+			}
+			if (Tstart[var] != (uint32_t)-1) {
+				op_live_total += liveliness_kill_var(op_array, cur_op, var, Tstart, opTs);
+			}
+		}
+	}
+
+#if ZEND_DEBUG
+	/* Check that all TMP variable live-ranges are closed */
+	for (i = 0; i < op_array->T; i++) {
+		ZEND_ASSERT(Tstart[i] == (uint32_t)-1);
+	}
+#endif
+
+	if (!op_live_total) {
+		info = NULL;
+	} else {
+		info = emalloc((op_array->last + 1 + op_live_total) * sizeof(uint32_t));
+
+		for (i = 0; i < op_array->last; i++) {
+			if (!opTs[i]) {
+				info[i] = (uint32_t)-1;
+			} else if (i > 0 && opTs[i-1] == opTs[i]) {
+				info[i] = info[i-1];
+			} else {
+				op_var_info *opT = opTs[i];
+				info[i] = info_off;
+				while (opT) {
+					info[info_off++] = opT->var;
+					opT = opT->next;
+				}
+				info[info_off++] = (uint32_t)-1;
+			}
+		}
+		info[op_array->last] = info_off;
```

Comment:
We should be able to drop the `+ 1` in emalloc() and assignment to info[op_array->last] now as we don't compare against the info[op_num+1] anymore.